components/data-provider/src/structs/dates.rs

```diff
@@ -0,0 +1,124 @@
+// Date
+#[cfg(feature = "invariant")]
+use crate::prelude::*;
+
+/// Gets a locale-invariant default struct given a data key in this module's category.
+#[cfg(feature = "invariant")]
+pub(crate) fn get_invariant(data_key: &DataKey) -> Option<DataResponse<'static>> {
+    use crate::invariant::make_inv_response;
+    if data_key.category != DataCategory::Dates {
+        return None;
+    }
+    match (data_key.sub_category.as_str(), data_key.version) {
+        ("gregory", 1) => make_inv_response::<gregory::DatesV1>(),
+        _ => None,
+    }
+}
+
+pub mod gregory {
+    use serde::{Deserialize, Serialize};
+    use std::borrow::Cow;
+    #[derive(Debug, PartialEq, Clone, Deserialize, Serialize, Default)]
+    pub struct DatesV1 {
+        pub symbols: DateSymbolsV1,
+
+        pub patterns: PatternsV1,
+    }
+
+    #[derive(Debug, PartialEq, Clone, Deserialize, Serialize, Default)]
+    pub struct DateSymbolsV1 {
+        pub months: months::ContextsV1,
+
+        pub weekdays: weekdays::ContextsV1,
+
+        pub day_periods: day_periods::ContextsV1,
+    }
+
+    #[derive(Debug, PartialEq, Clone, Deserialize, Serialize, Default)]
+    pub struct PatternsV1 {
+        pub date: patterns::StylePatternsV1,
+
+        pub time: patterns::StylePatternsV1,
+
+        pub date_time: patterns::StylePatternsV1,
+    }
+
+    macro_rules! symbols {
+        ($name: ident, $expr: ty) => {
+            pub mod $name {
+                use super::*;
+
+                #[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
+                pub struct SymbolsV1(pub $expr);
+
+                symbols!();
+            }
+        };
+        ($name: ident, $($element: ident: $ty: ty),*) => {
+            pub mod $name {
+                use super::*;
+
+                #[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
+                pub struct SymbolsV1 {
+                    $(pub $element: $ty),*
+                }
+                symbols!();
+            }
+        };
+        () => {
+            #[derive(Debug, PartialEq, Clone, Deserialize, Serialize, Default)]
+            pub struct WidthsV1 {
+                #[serde(skip_serializing_if = "Option::is_none")]
+                pub abbreviated: Option<SymbolsV1>,
+
+                #[serde(skip_serializing_if = "Option::is_none")]
+                pub narrow: Option<SymbolsV1>,
+
+                #[serde(skip_serializing_if = "Option::is_none")]
+                pub short: Option<SymbolsV1>,
```

Comment:
Seems like at least one entry in this struct should be non-optional, so that when the others are null, you fall back to the default (usually Short).

Reply:
I would assume that such validation happens on the struct production side, and/or consumption. I'm not sure how to elegantly express that constrain in the struct itself.
One way would be to make `short` mandatory, but that seems like a pretty heavy-handed `one-off` that won't work if the constrain is not hard on CLDR level.

So, I suggest we keep it as options and consider adding a check in the CldrJSONDataProvider to make sure that at least one is available.

Reply:
I guess what I'm saying is, at some point or another, we need to perform fallback or fail.  If I'm a user of ICU4X and I request a short date string, and the data is not available, do I get an error back out?

I don't think making `short` mandatory is heavy-handed.  CLDR's constraints are irrelevant when making this struct.  The goal of this struct should be to make Intl.DateTimeFormat as easy to implement as possible, and having null data with no fallback doesn't help toward that goal.

Reply:
Filed https://github.com/unicode-org/icu4x/issues/259 to discuss fallbacks.